mp/15 lean shift subst

mpenciak opened the pull request:
work towards #15 + generator for random Yatima Exprs for debugging

johnchandlerburnham commented on Tests/ExprGen.lean:
In order to write an `eval` function you would need a negative shift, since in a redex of the form `(lambda x => M) N` you need to decrement all the free variables in `M` when you do `M[x := N]`

mpenciak replied:
Will it ever need to shift back by more than 1? I deal with the -1 shift with `shiftBack` in the `subst` function, but can do it more generally

johnchandlerburnham replied:
Well, if we were doing `subst` for evaluation "for real" we would probably want to do some tricky combination of substitutions whenever we had application telescopes of `(lambda x y z => M) N1 N2 N3`

Since we're not going to use `subst` for this, I think `shiftBack` is probably fine, although I think I would prefer if it shifted `Var 0 ~> Var 0` or returned a `Maybe` rather than having an `unreachable!` which can panic

gabriel-barrett replied:
Why are we writing a substitution based eval/reduction? Is there any particular reason for this?

johnchandlerburnham replied:
We're not writing eval per se, but `subst` is a useful utility to have for various term manipulation things we might do it the parser. For example, I think we need it to convert inductive declarations into both inductive and constructor constants

johnchandlerburnham commented on Yatima/Expr.lean:
```diff
@@ -59,4 +59,67 @@ namespace Yatima
+def numBinders : Expr → Nat
+  | lam _ _ _ body => 1 + numBinders body
+  | pi _ _ _ body => 1 + numBinders body
+  | _ => 0
```
`letE` and `fix` are both binders

johnchandlerburnham commented on Yatima/Expr.lean:
```diff
@@ -59,4 +59,67 @@ namespace Yatima
+    pi name bind (type.substAux input subDep.succ) (body.substAux (.shift input 1 <| pure 0) subDep.succ) 
+  | letE name type value body => 
+    letE name (type.substAux input subDep) (value.substAux input subDep) (body.substAux (.shift input 1 <| pure 0) subDep.succ) 
+  | fix name body => sorry
```
`fix` is a binder like `lam`

johnchandlerburnham commented on Yatima/Expr.lean:
```diff
@@ -59,4 +59,67 @@ namespace Yatima
+  | app func input' => app (shiftBack func) (shiftBack input')
+  | letE name type value body => letE name (shiftBack type) (shiftBack value) (shiftBack body)
+  | fix name body => sorry
+  | other => other
```
Missing `lam` and `pi`, no?

johnchandlerburnham commented on Yatima/Expr.lean:
```diff
@@ -59,4 +59,67 @@ namespace Yatima
+    | some dep => if idx < dep then var name idx else var name <| idx + inc
+    | none     => var name <| idx + inc
+  | app func input              => app (func.shift inc dep) (input.shift inc dep)
+  | lam name bind type body     => lam name bind type (body.shift inc <| dep.map .succ)
```
You also have to shift the vars in the type

johnchandlerburnham commented on Yatima/Expr.lean:
```diff
@@ -59,4 +59,67 @@ namespace Yatima
+  | lam name bind type body     => lam name bind type (body.shift inc <| dep.map .succ)
+  | pi name bind type body      => pi name bind type (body.shift inc <| dep.map .succ)
+  | letE name type value body =>  
+    letE name type value (body.shift inc <| dep.map .succ)
```
and you have to shift the vars in the value

johnchandlerburnham commented:
Also requesting a review from @gabriel-barrett, who might be able to add some insight

gabriel-barrett commented:
It seems like this is overly complicated...
Basically, `subst expr input subDep` recurse down `expr`'s children nodes, but increasing `subDep` and shifting `input` by one whenever it finds a node that has a binding, namely, the body of `lam`, `fix`, `letE` and `pi` (only the "body" though, not the other nodes). The other nodes will recurse with `input` and `subDep` intact. And whenever it finds a variable, it can do three things:
- if the index of the variable is equal to `subDep` you return `input`
- if the index is less than `subDep` you return the variable as is
- otherwise you return a variable with the index-1

Now, the way I defined is quite inneficient, because it will shift `input`, traversing the whole `input`, every time you go through a binding node. Rather, the more efficient version is saving a `inc` variable which will increase as you recurse on binding nodes, and then at the last moment, i.e., when you would have to return `input`, when the index of the variable is equal to `subDep`, you return the shifted version of `input` using the `inc` variable.

Now, `shift` shouldn't take an `Option Nat`. The way it's written, `shift` is basically replicating the expression when you give it `Option.none`, and shifting by `idx` when you give it `Option.some idx`. I don't see why it would ever need to shift with `none`

johnchandlerburnham commented:
@gabriel-barrett The reason for the `Option` in shift is because in the parser we sometimes need to shift *all* variables by an increment, like in multiple binders `(x y z: A)`. The `none` case is supposed to do that

mpenciak commented:
I believe I addressed the comments on the `shift` and `subst` implementations. I followed the descriptions given by @gabriel-barrett  and what I found in Pierce's _Types and Programming Languages_ book.

My only question is about whether the shift should be occuring in the `types` for `lam` and `pi` expressions.

Finally, I also finished up a generator for random lambda expressions.

gabriel-barrett commented:
@johnchandlerburnham I didn't notice shifting with none had this effect. Shouldn't we write two functions then, one for shifting all variables, the other for shifting only the free variables, for efficiency? It seems to bit a bit inefficient matching some/none every time you recurse on a node. It's also a bit more clean having two functions, `shiftFreeVars` and `shiftAllVars` or something like that

gabriel-barrett commented:
@mpenciak it's almost correct. In the `var` case, you also have to check whether `idx > dep+acc`, in which case you return `var name (idx-1)`. The reason for this is that we don't want to leave a "hole" where the free variable given by de Bruijn *level* "dep" was; i.e., if the free variable that was substituted was 4, then 5 should take the place of 4, 6 of 5, 7 of 6, and so on.

Oh, and I think you should also add another function for shifting all variables (which previously was done by the `Option.none` case), because John needs it for the parser
